sematic/plugins/building/docker_builder.py

```diff
@@ -339,30 +388,170 @@ def _find_build_config_files(script_path: str) -> List[str]:
     return [file_candidate] if os.path.isfile(file_candidate) else []
 
 
-def _build_image(build_config: BuildConfig, target: str) -> ImageURI:
+def _build_image(
+    target: str,
+    build_config: BuildConfig,
+    docker_client: docker.DockerClient,  # type: ignore
+) -> Tuple[Image, ImageURI]:
     """
     Builds the container image to use, according to the build configuration, and returns
     an `ImageURI` that identifies it.
 
+    Parameters
+    ----------
+    target: str
+        The path to the pipeline target to launch; the built image must support this
+        target's execution.
+    build_config: BuildConfig
+        The configuration that controls the image build.
+    docker_client: docker.DockerClient
+        The client to use for executing the operations.
+
+    Returns
+    -------
+    Tuple[Image, ImageURI]:
+        The build container image and a URI that identifies the image.
+
     Raises
     ------
     BuildError:
-        There was an error when executing the specified build script.
+        There was an error when building the image.
     BuildConfigurationError:
         There was an error when validating the specified build configuration.
+    SystemExit:
+        A subprocess exited with an unexpected code.
     """
-    if build_config.base_uri is not None:
-        # TODO: honor configurations in build_config.build
-        return build_config.base_uri
+    effective_base_uri = build_config.base_uri
 
-    if build_config.image_script is None:
-        # appease mypy
-        raise BuildConfigurationError(
-            "Exactly one of `image_script` and `base_uri` must be specified!"
+    if build_config.image_script is not None:
+        effective_base_uri = _execute_build_script(
+            target=target, image_script=build_config.image_script
+        )
+
+    # appease mypy
+    assert effective_base_uri is not None
+
+    return _build_image_from_base(
+        target=target,
+        effective_base_uri=effective_base_uri,
+        build_config=build_config,
+        docker_client=docker_client,
+    )
+
+
+def _build_image_from_base(
+    target: str,
+    effective_base_uri: ImageURI,
+    build_config: BuildConfig,
+    docker_client: docker.DockerClient,  # type: ignore
+) -> Tuple[Image, ImageURI]:
+    """
+    Builds the container image to use by adding layers to an existing base image.
+    """
+    logger.info("Building image starting from base: %s", effective_base_uri)
+
+    dockerfile_contents = _generate_dockerfile_contents(
+        base_uri=effective_base_uri,
+        source_build_config=build_config.build,
+        target=target if build_config.base_uri is not None else None,
+    )
+    logger.info("Using Dockerfile:\n%s", dockerfile_contents)
```

Comment:
Can we make this `debug`? We want to limit log volume for the happy path.

Reply:
Done.